models/demos/tg/llama3_70b/tt/llama_attention_galaxy.py

```diff
@@ -467,3 +491,194 @@ def attn_selfout(
         )
 
         return attn_output
+
+    def prefill_forward(
+        self,
+        xs,
+        rot_mats,
+        attn_masks,
+        user_id: int = 0,
+    ):
+        query_layer, key_layer, value_layer = self.prefill_attn_qkv(xs, rot_mats)
+        attn_outputs = self.prefill_attn_mqa(query_layer, key_layer, value_layer, attn_masks, user_id)
+        return self.prefill_attn_selfout(attn_outputs)
+
+    def prefill_attn_qkv(
+        self,
+        xs,
+        rot_mats,
+    ):
+        assert xs.shape[1] == 1, "batch must be 1"
+        assert xs.shape[2] % 128 == 0 and xs.shape[2] > 0, "Seqlen must be divisible by 128"
+        _, _, seq_len, _ = xs.shape
+
+        max_mm_seq_len = self.model_config["MAX_MM_SEQ_LEN"]
+        batch_dim = 1 if seq_len < max_mm_seq_len else seq_len // max_mm_seq_len  # Find the division factor
+
+        # TODO: add batch dim for looping along it in case of seq_len > max_mm_seq_len
+        xs = ttnn.reshape(xs, (1, batch_dim, seq_len // batch_dim, -1))
+
+        # Fused QKV
+        fused_query_key_value = ttnn.matmul(
+            xs,
+            self.qkv,
+            dtype=ttnn.bfloat16,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            compute_kernel_config=self.COMPUTE_KERNEL_QKV,
+        )
+
+        fused_query_key_value = tt_all_reduce(
+            fused_query_key_value, self.device_mesh, cluster_axis=1, num_links=2, memory_config=ttnn.DRAM_MEMORY_CONFIG
+        )
+
+        fused_query_key_value = ttnn.reshape(fused_query_key_value, (1, 1, seq_len, -1))
+
+        xs.deallocate(True)
+
+        (
+            query_layer,  # [bsz, n_local_heads, seq_len, head_dim]
+            key_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+            value_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+        ) = ttnn.experimental.nlp_create_qkv_heads(
+            fused_query_key_value,
+            num_heads=self.n_local_heads,
+            num_kv_heads=self.n_local_kv_heads,
+            transpose_k_heads=False,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        )
+
+        fused_query_key_value.deallocate(True)
+
+        # ROTARY EMBEDDINGS
+        # Q Rotary Embeddings
+        # query_layer: ttnn.Shape([1, 8, seq_len, 128]) -> [bsz, n_local_heads, seq_len, head_dim]
```

Comment:
Maybe use the naming here for the first shape too? So that it's clear that we're switching bsz and n_local_heads?

Reply:
Here it is indeed the commented shape as is. We are swapping later.